File: app/src/main/java/com/kelsos/mbrc/helper/VolumeChangeHelper.java
package com.kelsos.mbrc.helper;

import android.widget.SeekBar;

import com.jakewharton.rxrelay.PublishRelay;

import java.util.concurrent.TimeUnit;

import rx.functions.Action1;

public class VolumeChangeHelper implements SeekBar.OnSeekBarChangeListener {

  private boolean userChangingVolume;
  private Action1<Integer> action;
  private PublishRelay<Integer> volumeRelay = PublishRelay.create();

  public VolumeChangeHelper(Action1<Integer> action) {
    this.action = action;
    this.userChangingVolume = false;
    volumeRelay.throttleLast(600, TimeUnit.MILLISECONDS).subscribe(this::onVolumeChange);
  }

  public boolean isUserChangingVolume() {
    return userChangingVolume;
  }

  @Override
  public void onProgressChanged(SeekBar seekBar, int value, boolean fromUser) {
    if (fromUser) {
      volumeRelay.call(value);
    }
  }

  @Override
  public void onStartTrackingTouch(SeekBar seekBar) {
    userChangingVolume = true;
  }

  @Override
  public void onStopTrackingTouch(SeekBar seekBar) {
    userChangingVolume = false;
  }

  private void onVolumeChange(int change) {
    if (action != null) {
      action.call(change);
    }
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/helper/VolumeChangeHelper.kt
package com.kelsos.mbrc.helper

import android.widget.SeekBar
import com.jakewharton.rxrelay.PublishRelay
import rx.functions.Action1
import java.util.concurrent.TimeUnit

class VolumeChangeHelper(private val action: Action1<Int>?) : SeekBar.OnSeekBarChangeListener {

  var isUserChangingVolume: Boolean = false
    private set
  private val volumeRelay = PublishRelay.create<Int>()

  init {
    this.isUserChangingVolume = false
    volumeRelay.throttleLast(600, TimeUnit.MILLISECONDS).subscribe { this.onVolumeChange(it) }
  }

  override fun onProgressChanged(seekBar: SeekBar, value: Int, fromUser: Boolean) {
    if (fromUser) {
      volumeRelay.call(value)
    }
  }

  override fun onStartTrackingTouch(seekBar: SeekBar) {
    isUserChangingVolume = true
  }

  override fun onStopTrackingTouch(seekBar: SeekBar) {
    isUserChangingVolume = false
  }

  private fun onVolumeChange(change: Int) {
    action?.call(change)
  }
}
